### DISCUSSION TITLE
Exposing Prisma's internal validation functions

### DISCUSSION THREAD

**Discussion by Nazaire** [prisma/prisma#18111] [Q&A]
The schema validation done by Prisma before sending requests is great. I need to separate the validation from the execution of the query, is there any way I can do that? 

To explain my use case, i'm bulk inserting 1000's of records, and if one fails then the entire transaction will fail too. I need to run the schema validation prior to invoking the request so I can filter out the invalid records. 

I've been looking around for a internal function that I can expose with `yarn patch` so I can call it when I need it. I can't find any information in the docs for something like this. 

Cheers, I appreciate the help .

**Comment by nurul3101** [comment]
Hello @Nazaire 👋 

You can use [Prisma Client Extensions](https://www.prisma.io/docs/concepts/components/prisma-client/client-extensions) to create a function that validates your data. You can invoke this function first and then invoke the insert query. Here's a reference [example](https://github.com/prisma/prisma-client-extensions/tree/main/input-validation) for input validation that you can use.

**Comment by Nazaire** [comment]
Hey @nurul3101 thanks for your example on how to add your own validation rules.

Prisma does it's own validation under the hood before submitting a query (i'm assuming based on the error output). For my case, **Prisma's internal validation is all I need**, I don't need to extend it just yet. 

How would I run Prisma's validation checks before invoking the query itself?

**Comment by Nazaire** [comment]
Let me elaborate with an example.

```
const objects = [ /* ... */ ]

// currently i'm inserting them all at once

prisma.model.createMany(objects)

// If there's an invalid object, this entire query will fail.

```

And what i'm asking for is:

```
const objects = [ /* ... */ ]
const validObjects = [];
for (const o of objects) {
     try {
         // invoke Prisma's validation function  <--- This is what my question is referring to
         Prisma.model.validate(o);
         validObjects.push(o);
     } catch (error) {
        // invalid schema
     }
}

// only insert the objects that passed validation

prisma.model.createMany(validObjects)

```

**Reply by nurul3101** [reply]
Are you using TypeScript? You can give the objects array a type of your `model`. The types for all the models are created when invoking `npx prisma generate` command. You can import the type from `"@prisma/client"`. It could look something like this
```ts
import { modelType } from "@prisma/client"

const objects: modelType = [ /* ... */ ]
```
If there are any objects that do not adhere to the types then you will get an error instantly before running the code.

**Reply by Nazaire** [reply]
Yes, i'm using Typescript. I am asking about the runtime validation.

**Comment by ludralph** [comment]
Hi there,

To keep our discussions organized and focused on the most relevant topics, we’re reviewing and tidying up our backlog. As part of this process, we’re closing discussions that haven’t had any recent activity and appear to be outdated.

If this discussion is still important to you or unresolved, we’d love to hear from you! Feel free to reopen it or start a new one with updated details.

For more details about our priorities and vision for the future of Prisma ORM, check out our latest blog post: https://www.prisma.io/blog/prisma-orm-manifesto.

Thank you for your understanding and being part of the community!
